File: macos/Scanner/Sources/Scanner/Rules/AESMCP012.swift
public struct AESMCP012: Rule {
    public let id = "AES-MCP-012"
    public let severity: Severity = .critical

    public init() {}

    public func evaluate(_ facts: ScanFacts) -> [Finding] {
        facts.mcpServers.compactMap { server in
            guard !server.disabled else {
                return nil
            }
            let reachesDockerDaemon = server.searchableValues.contains { value in
                AiThreatPatterns.grantsDockerDaemonAccess(value)
            }
            guard reachesDockerDaemon else {
                return nil
            }
            return finding(app: server.appId, serverName: server.name, affectedPath: server.configPath)
        }
    }
}

File: windows/Scanner/Rules/AESMCP012.cs
using AIExposureScanner.Scanner.Models;

namespace AIExposureScanner.Scanner.Rules;

public sealed class AESMCP012 : IRule
{
    public string Id => "AES-MCP-012";
    public Severity Severity => Severity.Critical;

    public IReadOnlyList<Finding> Evaluate(ScanFacts facts) =>
        facts.McpServers
            .Where(server => !server.Disabled)
            .Where(server => server.SearchableValues().Any(AiThreatPatterns.GrantsDockerDaemonAccess))
            .Select(server => this.Finding(server.AppId, server.Name, affectedPath: server.ConfigPath))
            .ToArray();
}
